src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java

```diff
@@ -85,11 +99,61 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
     protected abstract String parserName();
 
     protected abstract void doXContent(XContentBuilder builder, Params params) throws IOException;
-
     @Override
     public QueryValidationException validate() {
         // default impl does not validate, subclasses should override.
         //norelease to be removed once all queries support validation
         return null;
     }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        //norelease all querybuilders should finally implement their own streaming
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+      //norelease all querybuilders should finally implement their own streaming
+    }
+
+    protected static QueryBuilder readQueryBuilderFrom(StreamInput in) throws IOException {
+        QueryBuilder newBuilder;
+        String name = in.readString();
+        Class<?> newBuilderClass = queryBuilderRegistry.get(name);
+        if (newBuilderClass != null) {
+            try {
+                newBuilder = (QueryBuilder) newBuilderClass.newInstance();
```

Comment:
reflection... I'd try to avoid it if possible. 

Can we move from `Streamable` to `StreamableReader` for all query builders? `readFrom` would then create a new instance internally. In the registry we would just need empty objects, we don't need to create them all the time with reflection, just have a Map<String,QueryBuilder>, the lookup already returns an empty instance that we can use to call readFrom against, which will return the actual de-serialized object. Makes sense?